Simplenote/Snackbar.swift

```diff
@@ -0,0 +1,128 @@
+//
+//  Created by Kevin LaCoste on 2020-11-09.
+//  Copyright © 2020 Automattic. All rights reserved.
+//
+
+import Foundation
+
+class Snackbar {
+	
+	var message: String
+	
+	static var presenter = SnackbarPresenter()
+	
+	init(message: String = "Test message") {
+		self.message = message
+	}
+	
+	deinit {
+		print("Snackbar deinit")
+	}
+	
+	func show() {
+		print("Displaying snackbar")
+		Snackbar.presenter.present(self)
+	}
+}
+
+class SnackbarPresenter {
+	
+	var snackbar: Snackbar?
+	var viewController: SnackbarViewController?
+	
+	deinit {
+		print("SnackbarPresenter deinit")
+		// Should never happen
+	}
+	
+	func present(_ sender: Snackbar) {
+		guard snackbar == nil else {
+			print("Already presenting. Please wait.")
+			return
+		}
+		
+		snackbar = sender
+		
+//		let snackView = prepareView()
+		let snackView = prepareViewFromXIB()
+		presentView(snackView)
+		
+		// Puts the view onscreen and returns.
+	}
+	
+	func prepareView() -> UIView {
+		
+		// Prep a test view for display.
+		// Limit to 80% of the host window.
+		
+		let window = UIApplication.shared.keyWindow!
+		let width: CGFloat = window.frame.size.width * 0.8
+		let height: CGFloat = 60 // Should come from text size?
+		
+		let rect = CGRect(x: 0, y: 0, width: width, height: height)
+		let view = UIView(frame: rect)
+		view.backgroundColor = .lightGray
+		view.layer.cornerRadius = rect.height / 2
+		
+		return view
+	}
+	
+	func prepareViewFromXIB() -> UIView {
+		
+		let vc = SnackbarViewController()
+		let view = vc.view!
+		
+		let frame = view.frame
+		let newFrame = CGRect(x: 0, y: 0, width: frame.width - 30, height: frame.height)
+		vc.view.frame = newFrame
+		
+		viewController = vc
+		
+		return view
+	}
+
+	func presentView(_ view: UIView) {
+		
+		// Determine where the view goes in the host window.
+		
+		let window = UIApplication.shared.keyWindow!
+		
+		window.addSubview(view)
+		
+		// View center should be window width / 2 and window height - offset - (view height / 2)
+		
+		let viewHeight = view.frame.height
+		let offset: CGFloat = 50 // Magic number alert!
+		view.center.x = window.frame.size.width / 2
+		view.center.y = window.frame.size.height - offset - (viewHeight / 2.0) // Half the view height + the gap from view to screen edge.
+		print(view.center)
+		print(window.frame.height)
+		
+		// Now push the view down offscreen.
+		view.center.y = view.center.y + offset + viewHeight
+		print(view.center.y)
+		
+		UIView.animate(withDuration: 0.2, delay: 0.0, options: [], animations: {
+			view.center.y = view.center.y - offset - viewHeight
+		}) { _ in
+			print("Animation complete.")
+
+			DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
+				print("Hiding snackbar")
+
+				UIView.animate(withDuration: 0.66, delay: 0.0, options: []) {
+					view.alpha = 0
+				} completion: { (Bool) in
+					view.removeFromSuperview()
+					print("Done fade.")
+					
+					self.snackbar = nil
+					self.viewController = nil
+				}
+			}
+		}
+	}
```

Comment:
Few notes!

1. Please review encapsulation: It'd be neat if properties not meant for external usage were protected
2. Would it be possible to convert to **Autolayout**?
3. Please review Tabs / Spaces (the project itself uses Spaces rather than Tabs!)
4. Please use constants when appropriate (**UIKitConstants** may be of help!)
5. Rather than animating `asyncAfter` please consider using the `UIView.animate`'s **delay** parameter
6. What would happen if the Window rotates?
7. Please bear in mind that the action itself is optional

Reply:
Thank you for these! I was able to address some of them in my changes today with the exceptions being:

- 2. **Move to Auto Layout** — I haven't had a chance to work on this yet. I thought it would make sense to have the view properly configured (action vs no action) before digging into this. One question, do you recommend moving away from UIStackView as part of tackling this?
- 5. **Try without `asyncAfter`** — I tested this but using `UIView.animate`'s **delay** parameter caused the button to stop accepting input. Additionally, after wiring up the button, the code path changed a little to allow short-circuiting the timeout. If I've overlooked or misunderstood something here, please let me know and I'll have another look.
- 6. **Rotation** — Currently the Snackbar doesn't respond to rotation at all. I'm thinking this would be handled by adding some Auto Layout constraints to the view once the zoom-in animation completes but I haven't tried this yet.

I've updated the issue details to show the new API (new initializer) and added an updated GIF to show both styles in action. This still needs Auto Layout, theming for Dark/Light modes, and possibly a queuing mechanism if we want to allow queuing up of Snackbar notifications.